File: src/main/java/org/bsplines/ltexls/parsing/AnnotatedTextFragment.java
/* Copyright (C) 2020 Julian Valentin, LTeX Development Community
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package org.bsplines.ltexls.parsing;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.List;
import java.util.Map;
import org.bsplines.ltexls.server.LtexTextDocumentItem;
import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.languagetool.markup.AnnotatedText;
import org.languagetool.markup.AnnotatedTextBuilder;
import org.languagetool.markup.TextPart;

public class AnnotatedTextFragment {
  private AnnotatedText annotatedText;
  private CodeFragment codeFragment;
  private LtexTextDocumentItem document;
  private @MonotonicNonNull String plainText;
  private @MonotonicNonNull AnnotatedText inverseAnnotatedText;

  private @MonotonicNonNull Field annotatedTextMappingField;
  private @MonotonicNonNull Method mappingValueGetTotalPositionMethod;

  public AnnotatedTextFragment(AnnotatedText annotatedText, CodeFragment codeFragment,
        LtexTextDocumentItem document) {
    this.annotatedText = annotatedText;
    this.codeFragment = codeFragment;
    this.document = document;
  }

  public AnnotatedText getAnnotatedText() {
    return this.annotatedText;
  }

  public CodeFragment getCodeFragment() {
    return this.codeFragment;
  }

  public LtexTextDocumentItem getDocument() {
    return this.document;
  }

  public String getSubstringOfPlainText(int fromPos, int toPos) {
    if (this.plainText == null) this.plainText = this.annotatedText.getPlainText();

    if (this.inverseAnnotatedText == null) {
      AnnotatedTextBuilder inverseAnnotatedTextBuilder = new AnnotatedTextBuilder();
      List<TextPart> textParts = this.annotatedText.getParts();

      for (int i = 0; i < textParts.size(); i++) {
        TextPart textPart = textParts.get(i);

        switch (textPart.getType()) {
          case TEXT: {
            inverseAnnotatedTextBuilder.addText(textPart.getPart());
            break;
          }
          case MARKUP: {
            String markup;

            if ((i < textParts.size() - 1)
                  && (textParts.get(i + 1).getType() == TextPart.Type.FAKE_CONTENT)) {
              markup = textParts.get(i + 1).getPart();
              i++;
            } else {
              markup = "";
            }

            inverseAnnotatedTextBuilder.addMarkup(markup, textPart.getPart());
            break;
          }
          case FAKE_CONTENT: {
            inverseAnnotatedTextBuilder.addMarkup(textPart.getPart());
            break;
          }
          default: {
            break;
          }
        }
      }

      this.inverseAnnotatedText = inverseAnnotatedTextBuilder.build();
    }

    return this.plainText.substring(
        getOriginalTextPosition(this.inverseAnnotatedText, fromPos, false),
        getOriginalTextPosition(this.inverseAnnotatedText, toPos, true));
  }

  @SuppressWarnings({"unchecked", "rawtypes"})
  private int getOriginalTextPosition(AnnotatedText annotatedText, int plainTextPosition,
        boolean isToPos) {
    if (this.annotatedTextMappingField == null) {
      try {
        this.annotatedTextMappingField = annotatedText.getClass().getDeclaredField("mapping");
      } catch (NoSuchFieldException | SecurityException e) {
        throw new RuntimeException(e);
      }

      this.annotatedTextMappingField.setAccessible(true);
    }

    if (this.mappingValueGetTotalPositionMethod == null) {
      Class<?> mappingValueClass;

      try {
        mappingValueClass = Class.forName("org.languagetool.markup.MappingValue");
      } catch (ClassNotFoundException e) {
        throw new RuntimeException(e);
      }

      try {
        this.mappingValueGetTotalPositionMethod =
            mappingValueClass.getDeclaredMethod("getTotalPosition");
      } catch (NoSuchMethodException | SecurityException e) {
        throw new RuntimeException(e);
      }

      this.mappingValueGetTotalPositionMethod.setAccessible(true);
    }

    @NonNull Map<Integer, ?> mapping;

    try {
      @SuppressWarnings("assignment.type.incompatible")
      @NonNull Map<Integer, ?> mapping2 = (Map)this.annotatedTextMappingField.get(annotatedText);
      mapping = mapping2;
    } catch (IllegalArgumentException | IllegalAccessException e) {
      throw new RuntimeException(e);
    }

    for (Map.Entry<Integer, ?> entry : mapping.entrySet()) {
      if (entry.getKey() == plainTextPosition) {
        try {
          @SuppressWarnings("unboxing.of.nullable")
          int totalPosition = (int)this.mappingValueGetTotalPositionMethod.invoke(
              entry.getValue());
          return totalPosition;
        } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
          throw new RuntimeException(e);
        }
      }
    }

    return annotatedText.getOriginalTextPositionFor(plainTextPosition, isToPos);
  }
}

File: src/main/kotlin/org/bsplines/ltexls/parsing/AnnotatedTextFragment.kt
/* Copyright (C) 2019-2021 Julian Valentin, LTeX Development Community
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package org.bsplines.ltexls.parsing

import org.bsplines.ltexls.server.LtexTextDocumentItem
import org.bsplines.ltexls.tools.I18n
import org.bsplines.ltexls.tools.Logging
import org.languagetool.markup.AnnotatedText
import org.languagetool.markup.AnnotatedTextBuilder
import org.languagetool.markup.TextPart
import java.lang.reflect.Field
import java.lang.reflect.Method
import kotlin.math.roundToInt

class AnnotatedTextFragment(
  val annotatedText: AnnotatedText,
  val codeFragment: CodeFragment,
  val document: LtexTextDocumentItem,
) {
  private var plainText: String? = null
  private var inverseAnnotatedText: AnnotatedText? = null

  fun getSubstringOfPlainText(fromPos: Int, toPos: Int): String {
    val plainText: String = this.plainText ?: run {
      val plainText = this.annotatedText.plainText
      this.plainText = plainText
      plainText
    }
    val inverseAnnotatedText: AnnotatedText = this.inverseAnnotatedText ?: run {
      val inverseAnnotatedText: AnnotatedText = invertAnnotatedText(this.annotatedText)
      this.inverseAnnotatedText = inverseAnnotatedText
      inverseAnnotatedText
    }

    val plainTextFromPos: Int = getOriginalTextPosition(inverseAnnotatedText, fromPos, false)
    val plainTextToPos: Int = getOriginalTextPosition(inverseAnnotatedText, toPos, true)

    return if (plainTextFromPos <= plainTextToPos) {
      plainText.substring(plainTextFromPos, plainTextToPos)
    } else {
      Logging.logger.warning(
        I18n.format(
          "couldNotDeterminePlainTextPositions",
          fromPos,
          toPos,
          plainTextFromPos,
          plainTextToPos,
        )
      )
      ""
    }
  }

  companion object {
    private val annotatedTextMappingField: Field = run {
      val annotatedTextMappingField: Field = AnnotatedText::class.java.getDeclaredField("mapping")
      annotatedTextMappingField.isAccessible = true
      annotatedTextMappingField
    }

    private val mappingValueGetTotalPositionMethod: Method = run {
      val mappingValueClass: Class<*> = Class.forName("org.languagetool.markup.MappingValue")
      val mappingValueGetTotalPositionMethod: Method = mappingValueClass.getDeclaredMethod(
          "getTotalPosition")
      mappingValueGetTotalPositionMethod.isAccessible = true
      mappingValueGetTotalPositionMethod
    }

    fun getOriginalTextPosition(
      annotatedText: AnnotatedText,
      plainTextPos: Int,
      isToPos: Boolean,
    ): Int {
      @Suppress("UNCHECKED_CAST")
      val mapping: Map<Int, *> = annotatedTextMappingField.get(annotatedText) as Map<Int, *>

      val mappingList: MutableList<Pair<Int, Int>> = run {
        val mappingList = ArrayList<Pair<Int, Int>>()
        mappingList.add(Pair(0, 0))

        for ((key: Int, value: Any?) in mapping) {
          val totalPosition: Int = mappingValueGetTotalPositionMethod.invoke(value) as Int
          if (key == plainTextPos) return totalPosition
          mappingList.add(Pair(key, totalPosition))
        }

        mappingList
      }

      val i: Int = searchPlainTextPos(mappingList, plainTextPos, isToPos)
      val lowerNeighbor: Pair<Int, Int> = mappingList[i - 1]
      val upperNeighbor: Pair<Int, Int> = mappingList[i]

      if (!isToPos) {
        if (lowerNeighbor.first == plainTextPos) {
          return lowerNeighbor.second
        } else if (upperNeighbor.first == plainTextPos) {
          return upperNeighbor.second
        }
      } else {
        if (upperNeighbor.first == plainTextPos) {
          return upperNeighbor.second
        } else if (lowerNeighbor.first == plainTextPos) {
          return lowerNeighbor.second
        }
      }

      val t: Float = (
        (plainTextPos - lowerNeighbor.first).toFloat() /
        (upperNeighbor.first - lowerNeighbor.first).toFloat()
      )
      return ((1 - t) * lowerNeighbor.second + t * upperNeighbor.second).roundToInt()
    }

    private fun searchPlainTextPos(
      mappingList: MutableList<Pair<Int, Int>>,
      plainTextPos: Int,
      isToPos: Boolean,
    ): Int {
      // cannot use compareBy/thenBy, otherwise Jacoco tries to find a file "Comparisons.kt"
      // (probably due to inlining), which doesn't exist, and uploading to Coveralls fails
      val comparator = Comparator<Pair<Int, Int>> { a, b ->
        if (a.first < b.first) {
          -1
        } else if (a.first > b.first) {
          1
        } else if (a.second < b.second) {
          -1
        } else if (a.second > b.second) {
          1
        } else {
          0
        }
      }

      mappingList.sortWith(comparator)

      val pairToFind: Pair<Int, Int> = Pair(plainTextPos, Int.MAX_VALUE)
      var i: Int = -mappingList.binarySearch(pairToFind, comparator) - 1

      if (i <= 0) i = 1
      if (i >= mappingList.size) i = mappingList.size - 1

      if (isToPos) {
        while ((i < mappingList.size - 1) && (mappingList[i + 1].first <= plainTextPos)) i++
      } else {
        while ((i > 1) && (mappingList[i - 1].first >= plainTextPos)) i--
      }

      return i
    }

    fun invertAnnotatedText(annotatedText: AnnotatedText): AnnotatedText {
      val inverseAnnotatedTextBuilder = AnnotatedTextBuilder()
      val textParts: List<TextPart> = annotatedText.parts
      var i = 0

      while (i < textParts.size) {
        val textPart: TextPart = textParts[i]

        when (textPart.type) {
          TextPart.Type.TEXT -> inverseAnnotatedTextBuilder.addText(textPart.part)
          TextPart.Type.FAKE_CONTENT -> inverseAnnotatedTextBuilder.addMarkup(textPart.part)
          TextPart.Type.MARKUP -> {
            val markup: String = if (
              (i < textParts.size - 1) &&
              (textParts[i + 1].type == TextPart.Type.FAKE_CONTENT)
            ) {
              i++
              textParts[i].part
            } else {
              ""
            }

            inverseAnnotatedTextBuilder.addMarkup(markup, textPart.part)
          }
          null -> {}
        }

        i++
      }

      return inverseAnnotatedTextBuilder.build()
    }
  }
}
